### ISSUE TITLE
type[] can be used in a contravariant manner, which is unsound

### ISSUE THREAD

**Issue by Wizzerinus** [microsoft/pyright#11007]
**Describe the bug**
`type[X]` is covariant. However, unbound methods on `X` act in a contravariant manner. Hence, using such an unbound method causes unsound behavior.

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
from typing import reveal_type

class Foo:
    def foo(self) -> int:
        return 1

class Bar(Foo):
    def __init__(self, data: int):
        self.data = data

    def foo(self):
        return self.data

class Baz(Foo):
    def __init__(self, data: str):
        self.data = data

    def foo(self):
        return len(self.data)

def bad_func(typ: type[Foo], val: Foo) -> int:
    # typ.foo here is contravariant, which allows arbitrary class substitution from among any Foo's subclasses. If typ and val come from different subtyping branches, this is unsound.
    return typ.foo(val)

baz = Baz("something")
reveal_type(bad_func(Bar, baz))  # revealed: int, runtime: str
```

**VS Code extension or command-line**
CLI, version 1.1.405

**Comment by erictraut** [comment]
@carljm & @JelleZijlstra, I'm interested in your thoughts on this. None of the major Python type checkers (mypy, pyright, ty or pyrefly) detect an error in this case.

**Comment by carljm** [comment]
Yes, this is clearly unsound. The `self` argument is exempted from the usual contravariant checking of argument types in subclass override checks, because it is normally provided implicitly, but this exemption is unsound for using unbound methods from a `type[]` type.

I think in ty we'd want to at least provide an option to error on accessing unbound methods from a `type[]` type. Whether this could be opt-in or opt-out would depend on the observed ecosystem impact, I think.

**Comment by erictraut** [comment]
I'm going to change this from a bug report to an enhancement request. I think this would need to be added as a new diagnostic check. It could be quite noisy (many false positives) even in code that is strictly typed. I'm guessing that this is not a frequent source of error, so I consider this a low priority.

**Comment by JelleZijlstra** [comment]
JelleZijlstra/unsoundness#33 has been added to the collection. My intuition is that issuing a diagnostic for this would probably not be very noisy (it's rare to call instance methods on `type[]`), but it's not a high priority.
